Python/optimizer_bytecodes.c

```diff
@@ -407,6 +407,23 @@ dummy_func(void) {
             sym_set_type(value, &PyLong_Type);
             res = sym_new_type(ctx, &PyBool_Type);
         }
+        if (!sym_is_const(value)) {
+            assert(sym_matches_type(value, &PyLong_Type));
+            int next_opcode = (this_instr + 1)->opcode;
+            assert(next_opcode == _CHECK_VALIDITY_AND_SET_IP);
+            next_opcode = (this_instr + 2)->opcode;
+            // If the next uop is a guard, we can narrow value. However, we
+            // *can't* narrow res, since that would cause the guard to be
+            // removed and the narrowed value to be invalid:
+            if (next_opcode == _GUARD_IS_FALSE_POP) {
+                sym_set_const(value, Py_GetConstant(Py_CONSTANT_ZERO));
+                res = sym_new_type(ctx, &PyLong_Type);
```

Comment:
This is always a boolean, since it's the output of the `_TO_BOOL` op.

```suggestion
                res = sym_new_type(ctx, &PyBool_Type);
```